src/psfmachine/machine.py

```diff
@@ -1150,6 +1152,185 @@ def _get_mean_model(self):
+        mean_model_hd = Ap.dot(self.psf_w)
+        mean_model_hd[~np.isfinite(mean_model_hd)] = np.nan
+        mean_model_hd = mean_model_hd.reshape(phi_hd.shape)
+
+        # mask = (mean_model_hd >= -3) & (r_hd < 14)
+        # mean_model_hd_ma = mean_model_hd.copy()
+        # mean_model_hd_ma[~mask] = -np.inf
+
+        # mask out datapoint that don't contribuite to the psf
+        mean_model_hd_ma = mean_model_hd.copy()
+        mask = mean_model_hd > -3
+        mean_model_hd_ma[~mask] = -np.inf
+        mask &= ~((r_hd > 14) & (np.gradient(mean_model_hd_ma, axis=0) > 0))
+        mean_model_hd_ma[~mask] = -np.inf
+
+        # double integral using trapezoidal rule
+        integral = np.trapz(
+            np.trapz(10 ** mean_model_hd_ma, r_hd[:, 0], axis=0),
+            phi_hd[0, :],
+            axis=0,
+        )
+        # renormalize weights and build new shape model
+        if not self.normalized_shape_model:
+            self.psf_w *= np.log10(integral)
+            self.normalized_shape_model = True
+        self._get_mean_model()
+
+        if plot:
+            fig, ax = plt.subplots(1, 2, figsize=(9, 5))
+            im = ax[0].scatter(
+                phi_hd.ravel(),
+                r_hd.ravel(),
+                c=mean_model_hd_ma.ravel(),
+                vmin=-3,
+                vmax=-1,
+                s=1,
+                label=r"$\int = $" + f"{integral:.4f}",
+            )
+            im = ax[1].scatter(
+                r_hd.ravel() * np.cos(phi_hd.ravel()),
+                r_hd.ravel() * np.sin(phi_hd.ravel()),
+                c=mean_model_hd_ma.ravel(),
+                vmin=-3,
+                vmax=-1,
+                s=1,
+            )
+            ax[0].legend()
+            fig.colorbar(im, ax=ax, location="bottom")
+            plt.show()
+
+    def get_psf_metrics(self, resolution=10):
+        """
+        Computes three metrics for the PSF model:
+            source_psf_fraction: the amount of PSF in the data. Tells how much of a
+                sources is used to estimate the PSF, values are in between [0, 1].
+            perturbed_ratio_mean: the ratio between the mean model and perturbed model
+                for each source. Usefull to find when the time model affects the
+                mean value of the light curve.
+            perturbed_std: the standard deviation of the perturbed model for each
+                source. USeful to find when the time model introduces variability in the
+                light curve.
+
+        If resolution > 0, it creates high resolution shape models for each source by
+        dividing each pixels into a grid of [resolution x resolution]. This provides
+        a better estimate of `source_psf_fraction`.
+
+        Parameters
+        ----------
+        resolution : int
+            Value in which each pixel axis is split to increase resolution. Default is
+            0 for no subpixel resolution.
+
+        """
+        if resolution > 0:
```

Comment:
I think something can't have a "resolution" of zero. You might mean `npoints_per_pixel`? I think this is probably ok

Reply:
How long, roughly, does it take to calculate the high resolution metrics? Should we invest in doing this when we run on the supercomputer?

Reply:
It takes 5 sec to run for ~500 sources with 5400 pixels. 
I think `npoints_per_pixel` is a good name for the argument